### DISCUSSION TITLE
Config section for Export folders

### DISCUSSION THREAD

**Discussion by pustefisk** [KevinJump/uSync#688] [General]
Would it possible to make a Export Content config section so we set only the folders we would like to sync. As it is now it takes everything from root -> down. We need to be able select only folders we want to export!
If possible we would love to make a pull request for this change!?

**Comment by KevinJump** [comment]
Hi, 

It is possible to do some export restrictions in the config, (see https://docs.jumoo.co.uk/usync/uSync/reference/handlers#handler-specific-settings) so you can for example say don't export these folders or only export these bits, so this might do what you need on site wide bases. 

more granular control of exporting and importing via the UI is built into uSync.Exporter and uSync.Publisher, which are part of uSync.Complete (the paid bit). which let you pick bits of content as to act as the root of an export, or select what to push from a content node .

**Reply by pustefisk** [reply]
According to the code the ExportAllAsync allways take the root as start node (we are not using uSync.Complete):

virtual public async Task<IEnumerable<uSyncAction>>     ExportAllAsync(string[] folders, HandlerSettings settings, SyncUpdateCallback? callback)
    => await ExportAllAsync(default, folders, settings, callback);

Where the default param indicates no config is used for setting a start node.
Or am i missing something here ?

**Reply by KevinJump** [reply]
Hi.
In that instance the default is the parent (Container) so that call is passing to the one that says there is no parent. (so effectively root). 

The 'folders' in this instance are the on-disk folders and they mostly don't relate to the folders or structure you see inside Umbraco (they can a bit, but the default is for the folders structure to be flat, so often folders is just `./usync/path/content` ).

the settings are passed in the HandlerSettings  property which contains a number of things but include a Settings dictionary (which contains the exclude / include parameters). handlers can have a ShouldImport / ShouldExport method that then will be called on each element to work out if it should be imported or written out to disk. This is how the config does it. 

complete is playing a slightly different game, given a start node (or collection of nodes) and some dependency flags, it builds a dependent tree of things that you would want to export and then exports the individual items so that only the required things ever end up on the disk for the import potion, so you  don't actually get any extra files on disk you didn't ask for so the import/export checks are less relevant.

**Reply by pustefisk** [reply]
So what would the appsesstings.json uSync section look like if i want to exclude a folder and include another ?

**Reply by KevinJump** [reply]
Hi,

partial Example here (https://docs.jumoo.co.uk/usync/uSync/reference/handlers#example) which is for excluding bit in content, so if for example you hava stie structure like below. 

- Home 
   - About Us
   - Products
   - Contact

And you didn't want to import/export About Us. 

```json
"uSync" :{
    "Sets": {
        "Default": {
            "Handlers": {
                "ContentHandler" : {
                    "Settings" : {
                        "Exclude": "Home/AboutUs"
                    }
                }
            }
        }
    }
}
```

or if you wanted to *only* include Products, and Contact Us

```json
"uSync" :{
    "Sets": {
        "Default": {
            "Handlers": {
                "ContentHandler" : {
                    "Settings" : {
                        "Include": "Home/Products,Home/Contact"
                    }
                }
            }
        }
    }
}
```

under both settings below, things will still get exported to disk, they just won't get imported, if you also didn't want the files to end up on disk, you can add the `RulesOnExport` setting, and then the rules are checked both in and out (see https://docs.jumoo.co.uk/usync/uSync/reference/handlers#handler-specific-settings)

**Reply by pustefisk** [reply]
This is our Content - i only want to Export Demosite:
<img width="133" alt="image" src="https://github.com/user-attachments/assets/86a2ca78-95cf-40a9-b655-29158fca8e47">

It still Export everything. This is the config:

"uSync": {
  "Settings": {
    "RootFolder": "/uSync/v15/",
    "DefaultSet": "default",
    "ImportAtStartup": "none",
    "ExportAtStartup": "none",
    "ExportOnSave": "none",
    "ReportDebug": false,
    "AddOnPing": true,
    "CacheFolderKeys": true,
    "ShowVersionCheckWarning": false,
    "SignalRoot": "",
    "EnableHistory": true
  },
  "Sets": {
    "Default": {
      "Enabled": true,
      "DisabledHandlers": [
        //"contentHandler",
        "mediaHandler",
        "dictionaryHandler",
        "everythingHandler"
      ],
      "Handlers": {
        "ContentHandler": {
          "Settings": {
            "Include": "Demosite",
          }, 
          "RulesOnExport": true
        }
      }
    }
  }
},

Is anything wrong here ?

**Reply by KevinJump** [reply]
Hi, 

At a quick glance, rulesOnExport should be at the same level as the Include setting.

**Reply by pustefisk** [reply]
Right!
BUT - its still trying to export all content. It not in git changes but its still procesing all the files!
The config has a "Exclude" property but its still trying to export this part of the tree. What could be wrong ?

"Settings": {
  "ImportTrashed": false,
  "ExportTrashed": false,
  "Exclude": "Portal",
  "Include": "Demosite",
  "RulesOnExport": true
}
